src/mame/orca/vastar.cpp

```diff
@@ -411,28 +203,20 @@ void vastar_common_state::machine_start()
 
 void vastar_state::machine_reset()
 {
-	m_sprite_priority[0] = 0;
-
-	m_spriteram[0] = m_fgvideoram + 0x000;
-	m_bg_scroll[0] = m_fgvideoram + 0x3c0;
-	m_bg_scroll[1] = m_fgvideoram + 0x3e0;
-	m_spriteram[1] = m_fgvideoram + 0x400;
-	m_spriteram[2] = m_fgvideoram + 0x800;
 }
 
 WRITE_LINE_MEMBER(vastar_common_state::nmi_mask_w)
 {
 	m_nmi_mask = state;
 }
 
-
 void vastar_state::main_map(address_map &map)
 {
 	map(0x0000, 0x7fff).rom();
-	map(0x8000, 0x8fff).ram().w(FUNC(vastar_state::bgvideoram_w<1>)).share(m_bgvideoram[1]).mirror(0x2000);
-	map(0x9000, 0x9fff).ram().w(FUNC(vastar_state::bgvideoram_w<0>)).share(m_bgvideoram[0]).mirror(0x2000);
-	map(0xc000, 0xc000).writeonly().share(m_sprite_priority);   // sprite / BG priority
-	map(0xc400, 0xcfff).ram().w(FUNC(vastar_state::fgvideoram_w)).share(m_fgvideoram); // fg videoram + sprites
+	map(0x8000, 0x8fff).ram().w(m_vasvid, FUNC(vastar_video_device::bgvideoram1_w)).share("vasvid:bg1videoram").mirror(0x2000);
+	map(0x9000, 0x9fff).ram().w(m_vasvid, FUNC(vastar_video_device::bgvideoram0_w)).share("vasvid:bg0videoram").mirror(0x2000);
+	map(0xc000, 0xc000).w(m_vasvid, FUNC(vastar_video_device::priority_w));
+	map(0xc400, 0xcfff).ram().w(m_vasvid, FUNC(vastar_video_device::fgvideoram_w)).share("vasvid:fgvideoram");
```

Comment:
Is there a way to avoid the drivers needing to know they have to create shares with particular tags for the device?  It seems like a bit of a leaky abstraction when the drivers need to know implementation details.

Reply:
I don't know, is there? I just followed the pattern used by the existing video device used by the other games in the driver

Reply:
You can pass tags for any required_whatever<> object finder to a subdevice and they get their own reference to it.  Look at m_maincpu and m_rom in mame/apple/sonora.cpp/.h for an example of how that works on the device side.  You'd do it for required_share_ptrs rather than required_region_ptr or required_device.

Alternatively, if the device's exposed memory areas are always in the same order at the same relative offsets, you can have the device expose a map() (which sonora.cpp/.h also do) and then the driver doesn't even have to know about the shares, it's all encapsulated in the device.

Reply:
went with option A